Unify `ui` and `unit` Build & Cache Pipelines in CI

Hamms opened the pull request:
Follow-up to https://github.com/code-dot-org/code-dot-org/pull/65765, which updated our UI test pipeline in drone to use a simpler and more efficient design for building and caching the web application execution environment. This PR updates the unit test pipeline to use the same design, and removes a bunch of duplicated code in the process. We still *run* UI and unit tests in parallel in separate pipelines; those pipelines just now both use a single shared cache of build artifacts.

Currently our caching strategy for unit tests looks something like this (and our UI tests look significantly different):

- When someone merges a PR into staging:
  - Create a generic cache of the initial git checkout of staging latest
- When someone pushes commits to a feature branch:
  - Load the generic cache of the initial git checkout
  - Load a unit-test-specific cache of manually-selected build artifacts
  - Run all tests
  - If all tests pass, update the unit-test-specific cache with build artifacts from the current environment (note that this means the cache we loaded earlier is not a cache of staging, but rather a cache of whichever other feature branch happened to pass most recently)

With this PR, our caching strategy for both unit and UI tests will look something like:

- When someone merges a PR into staging:
  - Checkout staging latest from git
  - Run a full build
  - Cache the entire built project, minus a few (large) unnecessary directories
- When someone pushes commits to a feature branch:
  - Load the comprehensive "built staging" cache
  - Run all tests

This shifts a lot of the responsibility for maintaining our caching strategy out of the feature branches (which developers are often blocked by waiting on) and into the staging build (which most developers can safely ignore entirely).

type of run | before | after
--- | --- | ---
feature branch | <img width="270" height="343" alt="image" src="https://github.com/user-attachments/assets/8eff1785-8b4f-4e73-940f-fd1f583e9aac" /> | <img width="270" height="263" alt="image" src="https://github.com/user-attachments/assets/41b712f2-2bfd-494f-9db3-1f5d61c53def" />
merge to staging | <img width="270" height="223" alt="image" src="https://github.com/user-attachments/assets/f4289dd9-f621-46c1-b933-e063d6d805ba" /> | <img width="270" height="223" alt="image" src="https://github.com/user-attachments/assets/a40bc85e-d973-4c6a-8677-d3f90c784d06" />




Note one major implication of this change: **we are no longer performing separate builds for the UI and unit test pipelines**. We instead run a single generic reusable build with a single set of configuration options valid for either UI or unit tests. In this way, our CI tests will more closely resemble our CD tests. See https://github.com/code-dot-org/code-dot-org/pull/67771 for some of the changes which were necessary to make this possible.

And because we are no longer performing separate builds, we can also stop doing separate environment configuration! This PR combines our `prepare_unit_tests.sh`, `prepare_ui_tests.sh` and `prepare_tests_common.sh` config scripts into a single `prepare_ci_tests.sh` script.

## Testing story

Mostly relying on Drone itself to test this change to our Drone config.

Because the unit test pipeline attempts to only run tests against the subset of code which has changed from staging, I also relied heavily on [this branch](https://github.com/code-dot-org/code-dot-org/pull/67670) to do comparative performance testing against our current config.

## Deployment strategy

One unresolved question here is, "what kind of performance benefits will we get in practice when only running a subset of tests?" We should plan to keep an eye on Drone runtimes for a few days following deployment, to make sure this change isn't too optimized for the "run all tests" case.

## Follow-up work

By shifting more work out of the Drone builds that most developers see and interact with regularly and into the Drone builds that they do not, we run the risk of more errors going unnoticed. We should look into notification options for failed drone builds to find out if we can ping the infra team when staging builds fail.

We might also get value out of looking into ways of shrinking the cache file, or otherwise speeding up the `get-cached-staging-build` step.

## PR Creation Checklist:
<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy impacts have been documented
- [ ] Security impacts have been documented
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

Hamms committed:
merge unit and UI test prep steps into singular ci test prep

Hamms committed:
use new unified ci build in drone

Hamms committed:
export CI_JOB before running rake install

Hamms committed:
use robust DB setup in both UI and unit tests

Hamms committed:
Merge branch 'staging' into elijah/unify-drone-builds

Hamms committed:
actually seed UI tests in shared build

Hamms committed:
manage CI job in drone, so we can seed only when necessary

Hamms committed:
don't adjust db prep just yet

Hamms committed:
Merge branch 'staging' into elijah/unify-drone-builds

Hamms committed:
ensure some delicate tests continue to pass even with overridden hostnames

Hamms committed:
set CI_JOB in compose scripts like we do for .drone

Hamms committed:
attempt to use cached build

Hamms committed:
use renamed redis cluster in netsim, too

Hamms committed:
don't explicitly set hostnames; let lib/cdo.rb handle that programmatically

Hamms committed:
Merge branch 'staging' into elijah/unify-drone-builds

Hamms committed:
Revert "don't explicitly set hostnames; let lib/cdo.rb handle that programmatically"

This reverts commit d42797f8b4921b21ab52587507afd5746ae9990f.

Hamms committed:
attempt to connect securely to redis cluster in Drone tests, just to verify that it's not as simple as that to get it working

Hamms committed:
make unit test responsive to either secure or insecure environments, since we can't stub it

Hamms committed:
Revert "attempt to connect securely to redis cluster in Drone tests, just to verify that it's not as simple as that to get it working"

This reverts commit d0dc92ba94959c84a7f7303ead10d3000f3699d4.

Hamms committed:
tweak unit test to accommodate multiple different test environments

Hamms committed:
Merge branch 'staging' into elijah/unify-drone-builds

Hamms committed:
consistently stub all s3 directories in shared tests, not just sources

Hamms committed:
stub hostname in email renderer test, to guarantee consistency with static test fixtures

Hamms committed:
Merge branch 'staging' into elijah/unify-drone-builds

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
clean filesystem before building cache

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
remove a bunch more directories to see if it makes any difference at all

Hamms committed:
cleanup

Hamms committed:
don't unnecessarily optimize assets

Hamms committed:
Revert "don't unnecessarily optimize assets"

This reverts commit 85324894a27fdbac88238a90756d65bbef35fe71.

Hamms committed:
does retaining dashboard/tmp help speed up initialization?

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
yes it does; shaves about a minute off of the time it takes to precompile assets, but adds about a minute and a half to the time it takes to download the resulting cache file

This reverts commit 31ef2044463373473216ff1f19e0d3528695f1dd.

Hamms committed:
does retaining the yarn cache help speed up initialization?

Hamms committed:
Revert "does retaining the yarn cache help speed up initialization?"

This reverts commit a633ef6c59f4ce7dfaa4007e3c39de0a81127415.

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
remove temp changes

Hamms committed:
CORRECTLY remove temp changes

Hamms committed:
sign drone

Hamms committed:
Repair flaky test

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
temporarily tweak config to build a valid cache

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
remove some more cache directories so build doesn't break

Hamms committed:
don't remove nonexistant cache data

Hamms committed:
Revert "temporarily tweak config to build a valid cache"

This reverts commit 6105c5b0422a5e16c5a667c4c816b4e7410c0c8d.

Hamms committed:
re-sign drone

Hamms committed:
actually deflake redirects test; turns out it was failing whenever HoneybadgerTest executed before it did

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
re-sign drone

Hamms commented on .drone.yml:
```diff
@@ -319,25 +278,39 @@ steps:
+      - /entrypoint.sh docker/ci/scripts/prepare_ci_tests.sh
 
-  - name: cache-staging-misc-build-artifacts
+  # Optional optimization to reduce the size of the cache file we're about to
+  # generate. We manually delete a few large directories created by the build
+  # process which are not necessary to actually run the tests; local caches,
+  # logs, node_modules, etc.
+  - name: sanitize-cache
+    image: codedotorg/code-dot-org:1.16
+    commands:
+      - rm -r ./.venv/ # 214M
+      - rm -r ./apps/.yarn/ # 358M
+      - rm -r ./dashboard/tmp/ # 494M
+      - rm -r ./frontend/node_modules/ # 1.9G (!!!)
+      - rm -r ./frontend/.yarn/ # necessary to remove node_modules
+      - rm -r ./frontend/*/*/node_modules/ # ~115M
```
Super open to feedback on whether y'all think this optimization is worth the added complexity. The primary benefit here is speeding up the `get-cached-staging` operation; removing these directories (`frontend/node_modules` in particular) saves several minutes

Hamms commented on dashboard/test/integration/redirects_test.rb:
```diff
@@ -158,10 +158,15 @@ class RedirectsTest < ActionDispatch::IntegrationTest
-    # initialization time, so we must undo the stub that test_helper adds at
-    # runtime for this to resolve.
-    CDO.unstub(:override_dashboard)
+    # This route is defined based on the value of `CDO.dashboard_hostname`,
+    # which we alter in our test runs by stubbing out `CDO.override_dashboard`.
+    # We could theoretically unstub that value to attempt to mirror the
+    # configuration at the time we load the routes, but we also reload routes
+    # in other tests (see HoneybadgerTest in particular), so there's no
+    # guarantee which version will be loaded when this test executes.
+    #
+    # We must therefore reload the routes ourselves.
+    Rails.application.reload_routes!
```
This fix probably should have gone in https://github.com/code-dot-org/code-dot-org/pull/67771, but I didn't figure out what was going on here until after it was merged. Happy to pull this out into a standalone PR if y'all would like

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
re-sign drone after merge

Hamms commented on docker/ui-tests-compose.yml:
```diff
@@ -19,7 +19,6 @@
 # To clean up:
 # docker-compose -f ui-tests-compose.yml down
 
-version: "3"
```
This is just some misc cleanup while I was in the area, to resolve:

```
WARN[0000] /home/elijah/code-dot-org/docker/ui-tests-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion
```

davidsbailey commented on docker/ci/scripts/prepare_ci_tests.sh:
```diff
@@ -27,22 +50,21 @@ ignore_eyes_mismatches: true
 libraries_s3_directory: libraries_circle/$CI_BUILD_NUMBER
 sources_s3_directory: sources_circle/$CI_BUILD_NUMBER
-session_store_server: 'redis://ui-tests-redis:6379/0/session'
+geocoder_redis_url: 'redis://ci-tests-redis:6379/0/geocoder'
```
I don't think we want unit tests talking to redis. what do you think about blanking out this key in `dashboard/test/test_helper.rb`? I believe this is a direction that benefits us anyway, because it helps reduce the chances that developers will have local test failures due to the contents of their locals.yml.

Hamms replied:
I'm not sure I see the problem with unit tests talking to redis. Note that this is not changing that; unit tests are _currently_ talking to redis:

https://github.com/code-dot-org/code-dot-org/blob/6c46d4f3d5d46f8ae5f402e847083b63309b9e01/.drone.yml#L16-L20

https://github.com/code-dot-org/code-dot-org/blob/6c46d4f3d5d46f8ae5f402e847083b63309b9e01/docker/ci/scripts/prepare_unit_tests.sh#L32-L35

Can you clarify your concerns here?

davidsbailey replied:
good point -- geocoder_redis_url and session_store_server are not blocking for this PR since these values are unchanged for unit pipeline. however, it seems like taking away access to redis in unit tests would increase fault isolation between unit tests for anything related to session store and geocoder, because redis provides shared storage across unit tests.

Hamms replied:
I'm open to exploring making some changes to the way we think about redis in the context of tests, but let's do that in a follow-up. I'd like to keep this PR focused on preserving existing behavior as much as possible

davidsbailey replied:
great, I think we are on the same page here. I don't believe an immediate follow up is needed.

davidsbailey commented on docker/ci/scripts/prepare_ci_tests.sh:
```diff
@@ -27,22 +50,21 @@ ignore_eyes_mismatches: true
 no_https_store: true
 netsim_redis_groups:
-- master: redis://ui-tests-redis:6379
+- master: redis://ci-tests-redis:6379
```
unset in test_helper.rb?

Hamms replied:
I don't follow; can you clarify?

davidsbailey replied:
I have two general concerns with loosening restrictions on what unit tests can access:
- access to 3rd party services which can fail
- access to shared state which can leak between tests

I see we are using redis much like we use mysql (a internal service we provision), so that addresses my first concern. for state leakage, in the case of mysql, I understand how we are isolating unit tests from each other, but it's not clear to me whether/how that isolation is achieved for redis within unit tests. so, that makes me want to understand more about why we think it's safe to give netsim access to redis within unit tests.

if we don't want to have that conversation then an alternative would be to set these to their respective values via conditional logic, either here or in `config/test.yml.erb` or in `test_helper.rb`.

Hamms replied:
Great, thanks for the clarification! I'm aligned with you that those are valid concerns to be thinking about in general, but note that this PR does not actually represent a loosening of restrictions. Tests _already_ have access to redis; if we don't provide this config value, they will by default use a local instance:

https://github.com/code-dot-org/code-dot-org/blob/16dcb901136673db3e08ee7725e5e760c8c63727/dashboard/legacy/middleware/net_sim_api.rb#L34

https://github.com/code-dot-org/code-dot-org/blob/16dcb901136673db3e08ee7725e5e760c8c63727/dashboard/legacy/middleware/net_sim_api.rb#L432-L439

Note that this mirrors the pattern we use for geocoder and session store:

https://github.com/code-dot-org/code-dot-org/blob/16dcb901136673db3e08ee7725e5e760c8c63727/config.yml.erb#L280

https://github.com/code-dot-org/code-dot-org/blob/16dcb901136673db3e08ee7725e5e760c8c63727/dashboard/config/initializers/session_store.rb#L7

The change represented by this PR is much more innocuous than it seems; it represents a shift from pointing at local redis to pointing at an isolated redis service run in a separate container, to more closely adhere to Docker and Drone best practices.

In general, this PR is making minimal changes if any to the way in which our tests execute. It mostly changes the way we cache the work we do to prepare the environment in which the tests will execute. See https://github.com/code-dot-org/code-dot-org/pull/67771 for all of the changes that more directly impact the tests themselves.

I'm down to have a conversation about whether we want to broadly rethink the role redis plays within our application config, but I'd say that's out of scope for this PR.

davidsbailey replied:
ok, thanks for the additional info! this approach sounds good 👍

davidsbailey commented on docker/ci/scripts/prepare_unit_tests.sh:
```diff
@@ -1,43 +0,0 @@
-# Unit test settings and secrets
-build_i18n: true
-geocoder_redis_url: 'redis://unit-tests-redis:6379/0/geocoder'
-optimize_rails_assets: false
-optimize_webpack_assets: false
```
These lines seem important to restore. They could be set just for unit tests via new logic in test.yml.erb .

davidsbailey commented on docker/ci/scripts/prepare_unit_tests.sh:
```diff
@@ -1,43 +0,0 @@
-ignore_eyes_mismatches: true
-localize_apps: true
-use_my_apps: true
-skip_seed_all: true
```
how does removing this affect run time for unit test setup? I don't see any reason we'd want to seed all for unit tests.

Hamms replied:
Fortunately, we *also* don't have a reason to seed all for UI tests! This remains skipped in the new `prepare_ci_tests.sh`

https://github.com/code-dot-org/code-dot-org/blob/ac6652c15b9ddc4981e2f29197d288583f5aed88/docker/ci/scripts/prepare_ci_tests.sh#L55

davidsbailey reviewed (commented):
excited to see this consolidated! a few concerns inline about performance and any potential reduced restrictions on what services unit tests can access.

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
continue to use unoptimized assets in CI unit tests, but implement the exception in test.yml.erb rather than in drone config

Hamms committed:
deduplicate drone pipelines (#68844)

* stop caching ~/.rbenv in drone now that the gems are installed to the working directory

* temporarily build cache on PR

* apply bundle deployment mode to new drone config

* set deployment mode more consistently in CI build

* fix inconsistent identifiers in drone config

* undo temp changes and re-sign drone

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
fix typo

Hamms committed:
drop temporary naming scheme now that we've tested the relevant functionality

Hamms committed:
temporarily build cache on PR

Hamms committed:
Revert "temporarily build cache on PR"

This reverts commit 377b305c7682d7b51e1c8b22912f8fad901914e7.

Hamms committed:
re-sign drone

Hamms commented:
After lots of randomization, this PR is finally ready for re-review!

One notable change since last reviewed; I have incorporated [a tweak to how bundle installs gems](https://github.com/code-dot-org/code-dot-org/pull/68689). Specifically, we are now using the same `deployment` mode in CI as we use in CD. In an ideal world we would isolate that change from this caching change, but applying the bundle change before the caching change would involve messy temporary logic and applying it afterwards would result in our CI pipeline being in a temporarily fragile state. I think this approach is the lesser of the evils, but please let me know if you think there's a better way to handle it.

Otherwise, this PR remains focused on simplifying and deduplicating our caching strategy for creating reproducible Drone builds.

Thank you!

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms committed:
re-sign drone

Hamms committed:
Merge branch 'staging' into elijah/improve-unified-drone-build

Hamms commented:
Here's an example of the post-drone-instance-upsize effect this change will have on overall runtimes:

[before](https://drone.cdn-code.org/code-dot-org/code-dot-org/42148/2/6) | [after](https://drone.cdn-code.org/code-dot-org/code-dot-org/42147/2/5)
--- | ---
<img width="270" height="343" alt="image" src="https://github.com/user-attachments/assets/7ebee46b-df69-418e-bacf-6456a8caba0e" /> | <img width="270" height="263" alt="image" src="https://github.com/user-attachments/assets/a778699c-2441-4170-b854-6d21f676b4db" />
<img width="270" height="303" alt="image" src="https://github.com/user-attachments/assets/1b1a3063-752e-4453-b898-ad69ee5ab96f" /> | <img width="270" height="263" alt="image" src="https://github.com/user-attachments/assets/676cae77-a32d-4cd4-a5eb-4c59510e236d" />
